cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperator.java

```diff
@@ -119,12 +120,13 @@ public StatefulSetOperator(Vertx vertx, KubernetesClient client, long operationT
         return f;
     }
 
-    public Future<Void> maybeRestartPod(StatefulSet ss, String podName, boolean forceRestart) {
+    public Future<Void> maybeRestartPod(StatefulSet ss, String podName, Predicate<Pod> p) {
         long pollingIntervalMs = 1_000;
         long timeoutMs = operationTimeoutMs;
         String namespace = ss.getMetadata().getNamespace();
         String name = ss.getMetadata().getName();
-        if (isPodUpToDate(ss, podName) && !forceRestart) {
+        Pod pod = podOperations.get(ss.getMetadata().getNamespace(), podName);
+        if (isPodUpToDate(ss, podName) && !p.test(pod)) {
```

Comment:
We need to decide whether `isPodUpToDate()` should be a private concern of the SSO, or whether we should actually view it as just another predicate (and then chose to factor it into the KAO). Note that since the caller of `maybeRestartPod()` already has to pass a `SS` we don't even need to resort to `BiPredicate<StatefulSet, Pod>` to do this.

Reply:
My 2 cents on this decision ...
With this refactoring, we are moving the logic about the decision for a pod restart outside of the SS using a predicate. Having the pod updated or not in terms of generation is part of this logic. From a KAO perspective and who use the predicate, it seems that only the predicate drives the restart logic while leaving the `isPodUpToDate` internally there is a hidden logic. For this reason, my inclination is to move the `isPodUpToDate` in the KAO (so in the passed predicate). In this way, who reads the KAO code understand that the pod restart logic is just the predicate and there is no other hidden logic inside the SSO. Wdyt ?

Reply:
Yes, I agree it would be better to move this into KAO.